### ISSUE TITLE
The tenorflow_framework shared object contains public HWLOC symbols. which causes HWLOC unexpected behavior it links dynamically with some other library.

### ISSUE THREAD

**Issue by ivankochin** [tensorflow/tensorflow#40830]
**System information**
- Have I written custom code: No
- OS Platform and Distribution: Ubuntu 16.04, Fedora 30 (and possibly any other Linux)
- TensorFlow installed from: C binary package (libtensorflow-cpu-linux-x86_64-1.14.0.tar.gz)
- TensorFlow version (use command below): 1.14.0
- Python version: Not applicable

**Describe the current behavior**

When I run `nm libtensorflow_framework.so.1.14.0 | grep T | grep hwloc` on Linux, I get the following output:

```
...
00000000012d9fc0 T hwloc_bitmap_copy
00000000012d9fb0 T hwloc_bitmap_dup
00000000012da6e0 T hwloc_bitmap_fill
00000000012dbc20 T hwloc_bitmap_first
00000000012dbc90 T hwloc_bitmap_first_unset
00000000012d9ea0 T hwloc_bitmap_free
...
00000000012e78a0 T hwloc_topology_insert_group_object
00000000012e7a50 T hwloc_topology_insert_misc_object
00000000012e7110 T hwloc_topology_is_thissystem
00000000012e7b00 T hwloc_topology_load
00000000012e5bf0 T hwloc_topology_reconnect
00000000012e8330 T hwloc_topology_restrict
...
```
And many other publicly available HWLOC interfaces that are included in` tenorflow_framework.so` as public symbols.

So when some application linked with TensorFlow and some other library that dynamically links with HWLOC, then part of HWLOC interfaces calls located inside `libhwloc.so` but other interfaces located inside `tensorflow_framework.so` shared object. 

Libraries dependencies structure:

![tensorflow+hwloc](https://user-images.githubusercontent.com/22097249/85857265-2de8ee80-b7c2-11ea-823e-600e2840f4e6.png)

For example, when the library("Some Lib" on the scheme) call `hwloc_topology_init` then it goes to the HWLOC library located inside `tensorflow_framework.so`. But when it calls the `hwloc_topology_get_complete_cpuset` method, then it goes to the shared HWLOC library. This means that we have two instances of HWLOC library which are divided between two shared objects.

This situation causes HWLOC unexpected behavior inside other libraries (e.g. the "Some Lib" library on the scheme).

**Describe the expected behavior**

The expected behavior is to not deliver any HWLOC interfaces as part of public interface of the TensorFlow shared object.

**Standalone code to reproduce the issue:**

To reproduce this issue, we need at least two source code files: for the application and for the library.

some_lib.cpp:
```c++
#include <iostream>
#include <hwloc.h>

void print_bitmap(hwloc_const_bitmap_t bitmap) {
    if (bitmap == NULL) {
        printf("mask is: NULL\n");
        return;
    }
    char* buf = new char[256];
    hwloc_bitmap_snprintf(buf, 256, bitmap);
    printf("mask is: %s\n", buf);
    delete [] buf;
}

extern void print_topology_info() {
    hwloc_topology_t topology;
    hwloc_cpuset_t   process_cpu_affinity_mask;
    hwloc_nodeset_t  process_node_affinity_mask;

    // Parse topology
    hwloc_topology_init( &topology );
    hwloc_topology_load( topology );

    // Getting process affinity mask
    process_cpu_affinity_mask  = hwloc_bitmap_alloc();
    process_node_affinity_mask = hwloc_bitmap_alloc();

    hwloc_get_cpubind(topology, process_cpu_affinity_mask, 0);
    hwloc_cpuset_to_nodeset(topology, process_cpu_affinity_mask, process_node_affinity_mask);

    print_bitmap(process_cpu_affinity_mask);
    print_bitmap(process_node_affinity_mask);
    printf("Complete:\n");
    print_bitmap(hwloc_topology_get_complete_cpuset(topology));
    print_bitmap(hwloc_topology_get_complete_nodeset(topology));
    printf("Allowed:\n");
    print_bitmap(hwloc_topology_get_allowed_cpuset(topology));
    print_bitmap(hwloc_topology_get_allowed_nodeset(topology));
}
```

application.cpp:
```
#include <iostream>

extern void print_topology_info();

int main() {
    print_topology_info();
}
```

Build steps:
```
# another compiler is also applicable
g++ -c -fpic some_lib.cpp
g++ -shared -o some_lib.so some_lib.o -lhwloc

g++ application.cpp -o some_app -ltensorflow some_lib.so
```
Output:
```
mask is: 0x000000ff
mask is: 0x0
Complete:
mask is: NULL
mask is: 0x00000001
Allowed:
mask is: 0x000000ff
mask is: 0x00000001
```

Expected output:
```
mask is: 0x000000ff
mask is: 0xf...f
Complete:
mask is: 0x000000ff
mask is: 0xf...f
Allowed:
mask is: 0x000000ff
mask is: 0xf...f
```

**Other info/logs** :

Proofs of different HWLOC interfaces location:

Here is some information that was obtained via GDB during application debugging:

Shared objects location:
```
0x00007fffecdf1430  0x00007fffece1869a - tbbbind -> hwloc
0x00007fffed750e80  0x00007fffee5971b4 - libtensorflow_framework.so.1 (static hwloc)
```
Method addresses (can be easily correlated with corresponding shared object):
```
0x7fffecdf66a0 <hwloc_topology_init> - tbbbind -> hwloc
0x7fffecdf80c0 <hwloc_topology_load> - tbbbind -> hwloc
0x7fffecdfe2a0 <hwloc_get_cpubind>   - tbbbind -> hwloc
0x7fffeef4d6c8 <hwloc_cpuset_to_nodeset(hwloc_topology_t, hwloc_const_cpuset_t, hwloc_nodeset_t)> - libtensorflow_framework.so.1 (static hwloc)
0x7fffeef4d644 <hwloc_topology_get_complete_cpuset(hwloc_topology_t)> - libtensorflow_framework.so.1 (static hwloc)
0x7fffeef4d686 <hwloc_topology_get_complete_nodeset(hwloc_topology_t)> - libtensorflow_framework.so.1 (static hwloc)
0x7fffeef4d665 <hwloc_topology_get_allowed_cpuset(hwloc_topology_t)> - libtensorflow_framework.so.1 (static hwloc)
0x7fffeef4d6a7 <hwloc_topology_get_allowed_nodeset(hwloc_topology_t)> - libtensorflow_framework.so.1 (static hwloc)
```

**Workaround:**
Add `-lhwloc` flag to application compilation string to directly link the HWLOC shared object. After this change, all HWLOC interfaces will be located within the HWLOC shared object.

**Comment by mihaimaruseac** [comment]
Can you try with master/2.2/1.15 instead of 1.14? 1.14 has been released in a somewhat broken state and is no longer in the support window

**Comment by gunan** [comment]
Paul was working on this before he left. He added hwloc dependency instead of libnuma.
@guptapriya do you know who works with tensorflow sig-networking, and multi-CPU efforts?

I think the main question is if we need the hwloc symbols exported by TF libraries to be used by ops/kernels, potentially collectives?

**Comment by ivankochin** [comment]
@mihaimaruseac looks like TensorFlow 1.15 does not contain such public symbols and I cannot reproduce this issue with this version of the TensorFlow library.

**Comment by mihaimaruseac** [comment]
Thank you @ivankochin for checking. Can your application use 1.15?

**Comment by ivankochin** [comment]
@mihaimaruseac ,

I am a member of the team that develops the library that links with HWLOC dynamically. In the scheme below our library may be associated with the "Some Lib" element. 

We have no dependency on the TensorFlow library, but one of our customers affected by this unexpected behavior with Tensorflow 1.14, and the customer isn't able to update the TensorFlow version now.

**Comment by mihaimaruseac** [comment]
Unfortunately we cannot update TF 1.14 either as it is outside the support window, per our patch policy.

**Comment by guptapriya** [comment]
@gunan - I think @dubey is works with the SIG networking, among others.

**Comment by gunan** [comment]
@dubey could you take a look at this issue? Should we be exporting hwloc symbols?

**Comment by bramandia** [comment]
Hi Gunan, I checked with Ayush, he is not deeply involved in the NUMA effort as well as SIG networking.

**Comment by gunan** [comment]
Goldie, Could you help us find the SIG networking contact for this issue?

**Comment by byronyi** [comment]
@gunan Hi Guhan, SIG-Networking lead here. It is an issue that has been fixed a long time ago (since 1.15 I think). See https://github.com/tensorflow/tensorflow/pull/29807 for more details.

**Comment by gunan** [comment]
Thank you very much for the information @byronyi And as far as I can see, the report is about 1.14. So, the solution is to use a newer version of TF.
